### PR TITLE
Prometheus receiver perf

### PR DESCRIPTION
**Pull request by gracewehner**
**Description:**
Extends the testbed for perf testing for prometheus scraping. 
Tests are similar to TestMetric10kDPS scenarios with a couple changes due to the nature of scraping:
- Data items sent are per scrape interval (default 1s so all other tests are still the same)
- The batch label isn't added to the metric to send since this would create new time series every interval, which doubles the items to scrape every interval.
- Records the timestamp of a metric from each scrape to make sure the prometheus receiver is keeping up with the specified scrape interval.

Right now the test is prometheus receiver -> otlp exporter to test the receiver alone. Another test can be added for prometheus receiver -> prometheus exporter.

**Link to tracking Issue:**
https://github.com/open-telemetry/wg-prometheus/issues/19

**Documentation:**
Adds descriptions for prometheus sender and receiver and optional environment variables.

### CODE REVIEW

**Review by dashpole — APPROVED** [review]

**Review by vishiy — APPROVED** [review]

File: testbed/README.md [hunk 1]
```diff
@@ -29,11 +29,13 @@ For instance, if using the existing end-to-end test, the general dataflow can be
   * `OCMetricsDataSender` - Implementation of `DataSender` which sends to `opencensus` receiver.
   * `OTLPTraceDataSender` - Implementation of `DataSender` which sends to `otlp` receiver.
   * `OTLPMetricsDataSender` - Implementation of `DataSender` which sends to `otlp` receiver.
+  * `PrometheusDataSender` - Implementation of `DataSender ` which send sto `prometheus` receiver.
```
Comment (Aneurysm9) [inline testbed/README.md:32]:
```suggestion
  * `PrometheusDataSender` - Implementation of `DataSender ` which sends to `prometheus` receiver.
```

File: testbed/README.md [hunk 2]
```diff
@@ -29,11 +29,13 @@ For instance, if using the existing end-to-end test, the general dataflow can be
...
   * `ZipkinDataSender` - Implementation of `DataSender` which sends to `zipkin` receiver.
 * `DataReceiver` - Receives data from the collector instance under test and stores it for use in test assertions.
   * `OCDataReceiver` - Implementation of `DataReceiver` which receives data from `opencensus` exporter.
   * `JaegerDataReceiver` - Implementation of `DataReceiver` which receives data from `jaeger` exporter.
   * `OTLPDataReceiver` - Implementation of `DataReceiver` which receives data from `otlp` exporter.
+  * `PrometheusDataSender` - Implementation of `DataSender ` which receives data `prometheus` exporter.
```
Comment (Aneurysm9) [inline testbed/README.md:38]:
```suggestion
  * `PrometheusDataReceiver` - Implementation of `DataReceiver ` which receives data from `prometheus` exporter.
```

File: testbed/testbed/mock_backend.go [hunk 3]
```diff
@@ -148,6 +158,67 @@ func (mb *MockBackend) ConsumeMetric(md pdata.Metrics) {
 	if mb.isRecording {
 		mb.ReceivedMetrics = append(mb.ReceivedMetrics, md)
 	}
+	if mb.isScraping {
+		mb.getTimestamps(md)
+	}
+}
+
+func (mb *MockBackend) getTimestamps(md pdata.Metrics) {
```
Comment (Aneurysm9) [inline testbed/testbed/mock_backend.go:166]: I'm not terribly comfortable with this method having side effects that aren't apparent from its name.  Perhaps this could simply discover and return the `currentTimestamp` value and the last few lines could be performed instead in the `ConsumeMetric()` method where this one is invoked.

**Review by Aneurysm9 — APPROVED** [review]

**Review by vishiy — APPROVED** [review]

File: testbed/testbed/mock_backend.go (hunk 3 above)
Comment (gracewehner) [reply testbed/testbed/mock_backend.go]: Thanks, I have made this change

File: testbed/testbed/load_generator.go [hunk 4]
```diff
@@ -162,7 +173,7 @@ func (lg *LoadGenerator) generate() {
 
 		go func() {
 			defer workers.Done()
-			t := time.NewTicker(time.Second / time.Duration(lg.options.DataItemsPerSecond/lg.options.ItemsPerBatch/numWorkers))
+			t := time.NewTicker(lg.options.Interval / time.Duration(lg.options.DataItemsPerInterval/lg.options.ItemsPerBatch/numWorkers))
```
Comment (tigrannajaryan) [inline testbed/testbed/load_generator.go:176]: It is not clear why introduce options.Interval if that is not the actual interval between generations. The naming of options Interval/DataItemsPerInterval does not seem to accurately describe the behavior.
I think we can keep DataItemsPerSecond as is, and delete Interval which is not used anywhere else but this particular calculation. I don't see why the changes to options is necessary for this PR.

Comment (gracewehner) [reply testbed/testbed/load_generator.go:176]: Thanks @tigrannajaryan, I made this change to match the scrape interval chosen which Prometheus uses. I'll change it back to keep DataItemsPerSecond as is here and instead calculate DataItemsPerSecond based off the scrape interval in the Prometheus test

**Comment by bogdandrutu** [comment]
Please rebase

**Comment by alolita** [comment]
@gracewehner pl rebase and resubmit.

File: testbed/testbed/load_generator.go (hunk 4 above)
Comment (gracewehner) [reply testbed/testbed/load_generator.go]: @tigrannajaryan I have now pushed this change

File: testbed/testbed/data_providers.go [hunk 5]
```diff
@@ -130,10 +130,18 @@ func (dp *perfTestDataProvider) GenerateMetrics() (pdata.Metrics, bool) {
...
+			if dp.options.IsScraping {
+				labelsMap = map[string]string{
+					"item_index": "item_" + strconv.Itoa(j),
+				}
+			} else {
+				labelsMap = map[string]string{
+					"item_index":  "item_" + strconv.Itoa(j),
+					"batch_index": "batch_" + strconv.Itoa(int(batchIndex)),
+				}
+			}
```
Comment (tigrannajaryan) [inline testbed/testbed/data_providers.go:143]: Why is this change necessary?

File: testbed/testbed/mock_backend.go [hunk 6]
```diff
@@ -149,6 +159,67 @@ func (mb *MockBackend) ConsumeMetric(md pdata.Metrics) {
 	if mb.isRecording {
 		mb.ReceivedMetrics = append(mb.ReceivedMetrics, md)
 	}
+	if mb.isScraping {
+		currentTimestamp := getMetricTimestamp(md)
+		receivedTimestampsCount := len(mb.ReceivedTimestamps)
+		if receivedTimestampsCount == 0 || !mb.ReceivedTimestamps[receivedTimestampsCount-1].Equal(currentTimestamp) {
+			mb.ReceivedTimestamps = append(mb.ReceivedTimestamps, currentTimestamp)
+		}
+	}
```
Comment (tigrannajaryan) [inline testbed/testbed/mock_backend.go:168]: Please add comments to explain why when scraping this logic needs to be different and how it is different.

File: testbed/testbed/mock_backend.go [hunk 7]
```diff
@@ -149,6 +159,67 @@ func (mb *MockBackend) ConsumeMetric(md pdata.Metrics) {
...
+}
+
+func getMetricTimestamp(md pdata.Metrics) time.Time {
```
Comment (tigrannajaryan) [inline testbed/testbed/mock_backend.go:171]: Does this return the timestamp of the very first metric data point? If yes then getFirstMetricTimestamp may be a better name and some doc comment to explain what the func does would also be useful.

File: testbed/testbed/test_case.go [hunk 8]
```diff
@@ -169,7 +169,9 @@ func (tc *TestCase) StartAgent(args ...string) {
 	}()
 
 	endpoint := tc.LoadGenerator.sender.GetEndpoint()
-	if endpoint != nil {
+
+	// Prometheus endpoints use /metrics path instead of only tcp host and port
+	if endpoint != nil && tc.LoadGenerator.sender.ProtocolName() != "prometheus" {
```
Comment (tigrannajaryan) [inline testbed/testbed/test_case.go:174]: We should not be doing exceptions like this in the generic code. Please find a different way.

**Review by tigrannajaryan — CHANGES_REQUESTED** [review]

**Comment by punya** [comment]
@gracewehner would you mind addressing the problem found by the linter:

```
testbed/testbed/mock_backend.go:120:1: exported: comment on exported method MockBackend.EnableMetricTimestampRecording should be of the form "EnableMetricTimestampRecording ..." (revive)
// EnableTimestampRecording enables recording of metric timestamps by MockBackend.
^
```

Thanks!

File: testbed/testbed/data_providers.go (hunk 5 above)
Comment (gracewehner) [reply testbed/testbed/data_providers.go:146]: Including the batch index means adding new dimensions for the metric each time the data is sent. The Prometheus /metrics endpoint that's scraped will keep all timeseries, so the amount of data doubles each time new data is sent instead of keeping the same dimensions and a set amount of data to scrape. I have added a comment to explain this.

File: testbed/testbed/test_case.go (hunk 8 above)
Comment (gracewehner) [reply testbed/testbed/test_case.go:174]: This is tough since for the prometheus sender,`tc.LoadGenerator.sender.GetEndpoint()` is non-nil and is used to know what endpoint the sender is using to host the metrics, which is needed in the prometheus config of the prometheus receiver. However, the prometheus sender actually uses `<host>:<port>/metrics`, so the function `net.Dial(tc.LoadGenerator.sender.GetEndpoint().Network(), tc.LoadGenerator.sender.GetEndpoint().String())` errors since just the `<host>:<port>` endpoint isn't setup, and because of this error, the whole load test fails.

I have tried changing to have `endpointString` as a property of `PrometheusDataSender`, and not use the port and host of `DataSenderBase` but `GetEndpoint()` still returns `:0` since the `port` of type int defaults to 0. Should `GetEndpoint()` instead return nil if `Host` is an empty string and `Port` is 0?

**Review by gracewehner — COMMENTED** [review]
Addressing the comments. I will need to make another change since now the number of data items sent < the number of data items received because of the added internal prometheus metrics: up, scrape_duration_seconds, etc., so the validator fails the test.

**Review by tigrannajaryan — CHANGES_REQUESTED** [review]

**Comment by tigrannajaryan** [comment]
There is a still an issue with this PR: it contains Prometheus-specific code in generic testing-related files.

**Comment by alolita** [comment]
@gracewehner @vishiy can you please address @tigrannajaryan's concerns.

**Comment by alolita** [comment]
@gracewehner @vishiy can you please address @tigrannajaryan 's questions? Can you remove Prometheus-specific code from the general testing-related files.

**Comment by alolita** [comment]
@gracewehner can you rebase please?

File: testbed/testbed/mock_backend.go [hunk 9]
```diff
@@ -149,6 +159,59 @@ func (mb *MockBackend) ConsumeMetric(md pdata.Metrics) {
...
 		mb.ReceivedMetrics = append(mb.ReceivedMetrics, md)
 	}
+
+	// Record the timestamp of a metric from each scrape.
+	// Then the validator can check the difference between timestamps are the same as the scrape interval.
+	if mb.isRecordingMetricTimestamps {
+		currentTimestamp := getFirstMetricTimestamp(md)
+		receivedTimestampsCount := len(mb.ReceivedTimestamps)
+
+		// Record the timestamp only if it's the first recorded or if it's different from the last recorded timestamp
```
Comment (tigrannajaryan) [inline testbed/testbed/mock_backend.go:169]: Add a comment to explain why this is necessary. Why it matters that we only record the timestamps which are different from the last recorded. Why we record only the first timestamp if we receive several in this particular ConsumeMetric call.

File: testbed/testbed/senders.go [hunk 10]
```diff
@@ -84,6 +85,10 @@ type DataSenderBase struct {
 }
 
 func (dsb *DataSenderBase) GetEndpoint() net.Addr {
+	if dsb.Port == 0 && dsb.Host == "" {
+		return nil
+	}
```
Comment (tigrannajaryan) [inline testbed/testbed/senders.go:90]: Why is this check needed? Add a comment.

File: testbed/testbed/validator.go [hunk 11]
```diff
@@ -36,12 +36,43 @@ type TestCaseValidator interface {
 }
 
 // PerfTestValidator implements TestCaseValidator for test suites using PerformanceResults for summarizing results.
-type PerfTestValidator struct{}
+type PerfTestValidator struct {
+
+	// IsScraping specifies if the load generator
```
Comment (tigrannajaryan) [inline testbed/testbed/validator.go:41]: Incomplete sentence.

File: testbed/testbed/validator.go [hunk 12]
```diff
@@ -36,12 +36,43 @@ type TestCaseValidator interface {
...
+	IsScraping bool
+
+	// ScrapeInterval is the interval the receiver is scraping at, if IsScraping is enabled
+	ScrapeInterval time.Duration
+
+	// ExtraItemsScrapedPerIntervalCount accounts for if the receiver adds additional metrics than those sent
```
Comment (tigrannajaryan) [inline testbed/testbed/validator.go:47]: Unclear. I wonder if this is what is meant:
```suggestion
	// ExtraItemsScrapedPerIntervalCount accounts for if the receiver in the Collector adds additional metrics than those sent by the generator used by the testbed.
```

File: testbed/testbed/validator.go [hunk 13]
```diff
@@ -36,12 +36,43 @@ type TestCaseValidator interface {
...
+}
 
 func (v *PerfTestValidator) Validate(tc *TestCase) {
+	expectedDataItemsReceived := tc.MockBackend.DataItemsReceived()
+
+	if v.IsScraping {
+
+		// Get expected amount if additional metrics are added after scrape
+		expectedTotalExtraItemsReceived := (uint64(len(tc.MockBackend.ReceivedTimestamps) * v.ExtraItemsScrapedPerIntervalCount))
+		expectedDataItemsReceived = tc.MockBackend.DataItemsReceived() - expectedTotalExtraItemsReceived
```
Comment (tigrannajaryan) [inline testbed/testbed/validator.go:59]: Unclear what this calculation does, why the multiplication, then subtraction. Explain the logic in a comment.
I may be wrong but I think the names probably are not reflecting the intent, that may be contributing to the confusion. Perhaps use more descriptive names for variables.

File: testbed/tests/scenarios.go [hunk 14]
```diff
@@ -32,6 +36,7 @@ import (
 
 var (
 	performanceResultsSummary testbed.TestResultsSummary = &testbed.PerformanceResults{}
+	itemsPerIntervalEnvVar                               = "ITEMS_PER_INTERVAL"
```
Comment (tigrannajaryan) [inline testbed/tests/scenarios.go:39]: 1. This should be a `const` not `var`.
2. Why is this needed? Why is it more important to make this parameter available as en var, but not some other parameters, like for example data points per second?

File: testbed/tests/scenarios.go [hunk 15]
```diff
@@ -190,6 +195,89 @@ func Scenario10kItemsPerSecond(
...
+	resourceSpec testbed.ResourceSpec,
+	resultsSummary testbed.TestResultsSummary,
+	processors map[string]string,
+	extensions map[string]string,
+	scrapeInterval time.Duration,
+) {
+	resultDir, err := filepath.Abs(path.Join("results", t.Name()))
+	require.NoError(t, err)
+
+	// Because there are 7 datapoints per metric, this amounts to 10k items per second
```
Comment (tigrannajaryan) [inline testbed/tests/scenarios.go:211]: Where does the number 7 come from? If it is elsewhere how do we make sure the 2 places stay in sync?

File: testbed/tests/scenarios.go [hunk 16]
```diff
@@ -190,6 +195,89 @@ func Scenario10kItemsPerSecond(
...
+	}
+
+	dataProvider := testbed.NewPerfTestDataProvider(options)
+	tc := testbed.NewTestCase(
+		t,
+		dataProvider,
+		sender,
+		receiver,
+		agentProc,
+		&testbed.PerfTestValidator{IsScraping: true, ScrapeInterval: scrapeInterval, ExtraItemsScrapedPerIntervalCount: 5},
```
Comment (tigrannajaryan) [inline testbed/tests/scenarios.go:251]: Magic number 5. Where does this come from?

**Comment by bogdandrutu** [comment]
@gracewehner sorry, we moved the testbed code to contrib repository. Please file this PR there.
